src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -691,6 +692,307 @@ bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock
+            JITDUMP(" -- prev tree not ASG\n");
+            break;
+        }
+
+        GenTree* const prevTreeLHS = prevTree->AsOp()->gtOp1;
+        GenTree* const prevTreeRHS = prevTree->AsOp()->gtOp2;
+
+        if (!prevTreeLHS->OperIs(GT_LCL_VAR))
+        {
+            JITDUMP(" -- prev tree not ASG(LCL...)\n");
+            break;
+        }
+
+        // If we are seeing PHIs we have run out of interesting stmts.
+        //
+        if (prevTreeRHS->OperIs(GT_PHI))
+        {
+            JITDUMP(" -- prev tree is a phi\n");
+            break;
+        }
+
+        // Bail if RHS has an embedded assignment. We could handle this
+        // if we generalized the interference check we run below.
+        //
+        if ((prevTreeRHS->gtFlags & GTF_ASG) != 0)
+        {
+            JITDUMP(" -- prev tree RHS has embedded assignment\n");
+            break;
+        }
+
+        // If the VN of RHS is the VN of the current tree, or is "related", consider foward sub.
+        //
+        // Todo: generalize to allow when normal VN of RHS == normal VN of tree, and RHS except set contains tree except
+        // set.
+        // The concern here is whether can we forward sub an excepting (but otherwise non-side effecting tree) and
+        // guarantee
+        // the same side effect behavior.
+        //
+        // A common case we see is that there's a compare of an array element guarded by a bounds check,
+        // so prevTree is something like:
+        //
+        //    (ASG lcl, (COMMA (bds-check), (NE (array-indir), ...)))
+        //
+        // This may have the same normal VN as our tree, but also has an exception set.
+        //
+        // Another common pattern is that the prevTree contains a call.
+        //
+        const ValueNum domCmpVN    = prevTreeRHS->GetVN(VNK_Liberal);
+        const ValueNum domCmpSwpVN = vnStore->GetRelatedRelop(domCmpVN, ValueNumStore::VN_RELATION_KIND::VRK_Swap);
+        const ValueNum domCmpRevVN = vnStore->GetRelatedRelop(domCmpVN, ValueNumStore::VN_RELATION_KIND::VRK_Reverse);
+        const ValueNum domCmpSwpRevVN =
+            vnStore->GetRelatedRelop(domCmpVN, ValueNumStore::VN_RELATION_KIND::VRK_SwapReverse);
+
+        const bool matchCmp       = ((domCmpVN != ValueNumStore::NoVN) && (domCmpVN == treeVN));
+        const bool matchSwp       = ((domCmpSwpVN != ValueNumStore::NoVN) && (domCmpSwpVN == treeVN));
+        const bool matchRev       = ((domCmpRevVN != ValueNumStore::NoVN) && (domCmpRevVN == treeVN));
+        const bool matchSwpRev    = ((domCmpSwpRevVN != ValueNumStore::NoVN) && (domCmpSwpRevVN == treeVN));
+        const bool domIsSameRelop = matchCmp || matchSwp;
+        const bool domIsRevRelop  = matchRev || matchSwpRev;
+
+        // If the tree is some unrelated computation, keep looking farther up.
+        //
+        if (!domIsSameRelop && !domIsRevRelop)
+        {
+            JITDUMP(" -- prev tree VN is not related\n");
+            continue;
+        }
```

Comment:
I don't understand how this works. Doesn't this effectively make the interference check below useless? E.g.
```csharp
Foo(3, 4);

void Foo(int a, int b)
{
  bool foo = a < b;
  a = 15;
  if (foo)
    Console.WriteLine(a);
  else
    Console.WriteLine(b);
}
```
15 does not have the same VN as `a < b`, so the assignment to `a` will be skipped and the program will always print `4` when it should print `15`.

Reply:
Yeah, we should be adding `a` to the interfering set here before continuing the search.